### PR TITLE
[Snippets] Added Configurator support

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Added `RuntimeConfigurator` support to configure runtime arguments (loop parameters, data offsets) for each static shapes in dynamic case.*
- *Added Shape-Agnostic dynamism for Eltwise subgraphs (disabled by default)*

### Tickets:
 - *113211*
 - *139795*
 - *136157*
 - *139796*
 - *139785*

### Prerequisites:
- https://github.com/openvinotoolkit/openvino/pull/21934
- https://github.com/openvinotoolkit/openvino/pull/21881
- https://github.com/openvinotoolkit/openvino/pull/21895
- https://github.com/openvinotoolkit/openvino/pull/23710
- https://github.com/openvinotoolkit/openvino/pull/24115
- https://github.com/openvinotoolkit/openvino/pull/24119

### TODO:
- [x] Add `ShapeInfer` caching
- [x] Fix all failed tests

### CODE REVIEW

File: src/common/snippets/src/lowered/pass/optimize_domain.cpp [hunk 1]
```diff
@@ -78,12 +78,13 @@ inline bool OptimizeDomain::can_increase_jit_work_amount(const VectorDims& maste
 }
 bool OptimizeDomain::run(snippets::lowered::LinearIR& linear_ir) {
     OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::OptimizeDomain")
-    const auto& config = linear_ir.get_config();
-    if (linear_ir.empty())
+    // If m_tile_rank > 0, it means that backends manually set this value
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/optimize_domain.cpp:81]: This is not supposed to work this way 🙂
We invested resources into PassConfig, so the plugin can disable any transformation.
This seems like a perfect use case for this feature.

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 2]
```diff
@@ -69,6 +71,8 @@ std::shared_ptr<LinearIR> LinearIR::clone() const {
     // It's Ok to share shapeInfer factory ptr, since the factory doesn't depend on LIR in any way
     cloned->m_shape_infer_factory = m_shape_infer_factory;
     cloned->m_shape_infer = std::make_shared<LIRShapeInfer>(cloned->m_expressions, cloned->m_io_expressions);
+    cloned->m_runtime_configurator = m_runtime_configurator->clone(*cloned);
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/linear_ir.cpp:74]: It is strange to see that `clone()` take input arguments. If we want to clone an existing class instance, the instance already knows all the field values, so why do we need some arguments?
`clone()` with arguments looks like a constructor to me 🙂

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 3]
```diff
@@ -379,6 +383,14 @@ VectorDims LinearIR::get_master_shape() const {
     return master_shape;
 }
 
+void LinearIR::init_runtime_configurator() {
+    m_lowered_config = m_runtime_configurator->init(*this);
+}
+
+const RuntimeConfig& LinearIR::configure_runtime_args() {
+    return m_runtime_configurator->update(*this);
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/linear_ir.cpp:391]: Also discussed also in op::Subgraph: not sure why we need these bypass methods in LIR interface.
Look at the signatures of configurator's `init()` and `update()` methods: it takes LIR as an argument, and this makes perfect sense, since the configurator needs an IR to compute runtime properties. But the opposite is not valid, LIR doesn't need the configurator at all, it doesn't affect it's ability to maintain the graph state and apply transformations.

**Review by IvanNovoselov — COMMENTED** [review]
Ok, I partially reviewed the PR and left some high-level comments, but I intentionally didn't comment on the main parts: `runtime_config` and `runtime_configurator`. 
There are a lot of questions and concerns, but I don't think it is worth to comment on them, since I don't have the full picture without comprehensive use cases.
I think the review would be much more productive if you could present the design first, so we have a broader vision of the future application.

File: src/common/snippets/include/snippets/lowered/runtime_configurator.hpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    std::vector<PortDescriptorPtr> m_io_descs = {};
+    std::vector<size_t> m_io_data_sizes = {};
+
+    std::map<RuntimeConfig::LoopDescriptor::Type, std::shared_ptr<RuntimeConfigurator::LoopInitializer>> m_desc_initializers;
+    RuntimeConfig m_config;
+    bool m_inited = false;
+};
+
+
+class RuntimeConfigurator::LoopInitializer {
```
Comment (IvanNovoselov) [inline src/common/snippets/include/snippets/lowered/runtime_configurator.hpp:95]: We have a special class responsible for loop management `LinearIR::LoopManager`. It owns information about all loops, it has their loop_info, which in turn has all the information about the loop properties and handlers used to process specific iterations. 
The question is do we really need to introduce 4 additional classes just to copy a couple fields from loop_info to LoopDescriptor and occasionally call some `init_` functions (again, some updates based on loop_info)?
Can we instead create a ` RuntimeConfig::LoopDescriptor::update(const LoopInfoPtr& loop_info)` that will update the LoopDescriptor based on loop_info?
Another suggestion is to create `LoopDescriptor` from loop_info directly, and store the pointer to "bounded" loop_info there. In this case you will be able to call `update()` without any arguments, and it will do all the same work without extra fuss.

**Review by IvanNovoselov — COMMENTED** [review]
Up to LoopInitilizers. RuntimeConfig is not covered.

**Review by v-Golubev — COMMENTED** [review]
First part

File: src/common/snippets/include/snippets/lowered/expression.hpp [hunk 5]
```diff
@@ -104,6 +105,17 @@ class IOExpression : public Expression {
     io_type m_type = io_type::UNDEFINED;
 };
 
+class LoopExpression : public Expression {
```
Comment (a-sidorova) [inline src/common/snippets/include/snippets/lowered/expression.hpp:108]: Waiting for the https://github.com/openvinotoolkit/openvino/pull/24152

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 6]
```diff
@@ -32,23 +33,28 @@ class Config {
...
-#endif
     // Some Subgraphs doesn't support domain optimization due to operations' semantics
     bool m_enable_domain_optimization = false;
+    // True if the Buffer scratchpad size of LinearIR will be optimized (all possible optimizations will be activated)
+    // False if all Buffers will have uniqie ID and offsets in the Linear IR
+    bool m_are_buffers_optimized = true;
+    // Defined by OptimizeDomain optimization. Otherwise can be inited by backend
+    size_t m_loop_depth = 0;
+    // If it's zero (default value), data offsets and master shape won't be appended by ones. Can be inited by backed.
+    size_t m_tensor_rank = 0;
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp]: Removed - now it's not needed - `CPUConfigurator` inits the value itself

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 7]
```diff
@@ -253,6 +264,8 @@ class LinearIR {
     Config m_config{};
     LoopManagerPtr m_loop_manager = nullptr;
     std::shared_ptr<IShapeInferSnippetsFactory> m_shape_infer_factory;
+    std::shared_ptr<RuntimeConfigurator> m_runtime_configurator;
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp]: Now `RuntimeConfigurator` is part of `TargetMachine`

File: src/common/snippets/include/snippets/op/kernel.hpp [hunk 8]
```diff
@@ -32,8 +32,13 @@ class KernelStatic : public Kernel {
 public:
     OPENVINO_OP("KernelStatic", "SnippetsOpset", Kernel);
     KernelStatic() = default;
-    KernelStatic(lowered::LinearIR region);
+    KernelStatic(lowered::LinearIR region, std::vector<std::vector<size_t>> data_offsets);
     std::shared_ptr<Node> clone_with_new_inputs(const OutputVector& inputs) const override;
+
+    const std::vector<std::vector<size_t>>& get_data_offsets() const { return m_data_offsets; }
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/op/kernel.hpp]: Moved to `CPURuntimeConfigurator`. Also now `CPURuntimeConfigurator` returns `data_offsets` that can be set to `compile_args` in static pipeline

File: src/common/snippets/src/op/subgraph.cpp [hunk 9]
```diff
@@ -442,6 +460,15 @@ void Subgraph::control_flow_transformations(lowered::LinearIR& linear_ir,
...
     pipeline.run(linear_ir);
+
+    // After transformations init runtime configurator
+    // If shapes are static - creates static loop descriptors with inited data ptr shifts
+    // If shapes are dynamic - creates full set of loop descriptor for the following initialization by static shapes
+    linear_ir.init_runtime_configurator();
+}
+
+const lowered::RuntimeConfig& Subgraph::configure_runtime_args() {
+    return m_linear_ir->configure_runtime_args();
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/op/subgraph.cpp]: Discussed offline: configurator is stored in target machine however backend can configure the config using function calls from `Subgraph` to `TargetMachine`

File: src/common/snippets/src/lowered/pass/optimize_domain.cpp (hunk 1 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/pass/optimize_domain.cpp]: Fully agree. Removed

File: src/common/snippets/include/snippets/lowered/runtime_configurator.hpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+public:
+    RuntimeConfigurator();
+
+    /**
+     * @brief Initialize config using LinearIR state
+     * @param linear_ir the updated LinearIR
+     * @return config with runtime parameters
+     */
+    const RuntimeConfig& init(const lowered::LinearIR& linear_ir);
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/runtime_configurator.hpp]: Removed the method `init` at all: leave only one public method `get_updated_config(LIR)`

File: src/common/snippets/include/snippets/lowered/runtime_configurator.hpp [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    /**
+     * @brief Update config using LinearIR state
+     *        Note: Must be called only on static shapes!
+     * @param linear_ir the updated LinearIR
+     * @return config with runtime parameters
+     */
+    const RuntimeConfig& update(const lowered::LinearIR& linear_ir);
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/runtime_configurator.hpp]: Removed the method `init` at all: leave only one public method `get_updated_config(LIR)`

File: src/common/snippets/src/lowered/runtime_configurator.cpp [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    init_loop_descriptors(linear_ir.get_loop_manager());
+    init_data_offsets();
+    m_inited = true;
+    return m_config;
+}
+
+const RuntimeConfig& RuntimeConfigurator::update(const lowered::LinearIR& linear_ir) {
+    OPENVINO_ASSERT(m_inited, "Configurator must be already inited!");
+    update_loop_descriptors(linear_ir.get_loop_manager());
+    init_data_offsets();
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/lowered/runtime_configurator.cpp]: Added, thanks!

File: src/common/snippets/include/snippets/lowered/runtime_configurator.hpp (hunk 4 above)
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/runtime_configurator.hpp]: Removed `LoopDescriptor` and `LoopInitializer` classes at all.
Now we work with `ExpandedLoopInfo`

File: src/common/snippets/src/lowered/pass/update_loop_info.cpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            InitLoops::init_loop_info(updated_unified_loop_info, true);
+
+            current_work_amount = updated_unified_loop_info->get_work_amount();
+            init_data_ptr_shifts(updated_unified_loop_info, ptr_increments, finalization_offsets);
+        }
+
+        const auto& decomposed_loop_type = expanded_loop_info->get_type();
+
+        // If the specific iteration is not needed, we skip loop evaluation - set zero as work amount is enough
+        if (!InsertSpecificIterations::is_decomposed_loop_needed(updated_unified_loop_info, decomposed_loop_type, current_work_amount)) {
```
Comment (a-sidorova) [inline src/common/snippets/src/lowered/pass/update_loop_info.cpp:57]: There are the following ideas:
1. Move all decomposed loop (`ExpandedLoopInfo`) related static methods from `InsertSpecificIterations` pass to interface of `ExpandedLoopInfo`
2. Create new base class for `ExpandedLoopInfo` intialization that contains all decomposed loop related static methods and derive `InsertSpecificIterations` and `UpdateLoopInfo` from this class. One more idea: split `InsertSpecificIterations` into classes `InitExpandedLoopInfo` that returns `ExpandedLoopInfo` collection and `InsertSpecificIterations` that inserts new loop iterations using this collection

**Review by v-Golubev — COMMENTED** [review]
First part

File: src/plugins/intel_cpu/src/emitters/snippets/jit_snippets_call_args.hpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+#include <cstdint>
+#include <cstddef>
+
+#include "oneapi/dnnl/dnnl_common_types.h"
+
+namespace ov {
+namespace intel_cpu {
+
+#define SNIPPETS_MAX_SNIPPETS_DIMS 12
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/emitters/snippets/jit_snippets_call_args.hpp:16]: It's used below to set static array sizes. But I agree that the name is confused. Renamed. Thanks

[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/plugins/intel_cpu/src/emitters/snippets/jit_snippets_call_args.hpp]: Don't know why I missed this usage :) Thanks for the renaming, I think now it's better

**Review by v-Golubev — COMMENTED** [review]
A few more comments

File: src/common/snippets/include/snippets/op/subgraph.hpp [hunk 15]
```diff
@@ -144,17 +134,38 @@ class Subgraph : public ov::op::util::SubGraphOp {
     void data_flow_transformations(const BlockedShapeVector& blocked_input_shapes = {},
                                    const std::vector<ov::element::Type>& input_precisions = {},
                                    const std::vector<ov::element::Type>& output_precisions = {},
-                                   const std::vector<snippets::pass::Manager::PositionedPassBase>& = {});
+                                   const std::vector<snippets::pass::Manager::PositionedPassBase>& = {}) const;
+
+    void control_flow_transformations(const std::shared_ptr<lowered::pass::PassConfig>& lowered_pass_config = std::make_shared<lowered::pass::PassConfig>(),
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/op/subgraph.hpp]: This method should be private for now. I moved to private section since it can be called from `lowering_transformations`. Thanks

File: src/common/snippets/src/runtime_configurator.cpp [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+namespace ov {
+namespace snippets {
+
+RuntimeConfigurator::RuntimeConfigurator(std::shared_ptr<RuntimeConfig> c) : m_config(std::move(c)) {
+    OPENVINO_ASSERT(m_config, "Runtime config is nullptr!");
+
+    // Init LinearIR StateUpdater: some passes to update LoopInfo, BufferInfo etc
+    m_state_updater = lowered::pass::PassPipeline();
+    m_state_updater.register_pass<lowered::pass::UpdateLoopInfo>();
+}
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/runtime_configurator.cpp]: Done, thanks

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 17]
```diff
@@ -51,117 +51,187 @@ std::mutex err_print_lock;
...
+
+struct SubgraphShapeAgnosticKey : public SubgraphKey {
+    SubgraphShapeAgnosticKey(const std::shared_ptr<Subgraph::SubgraphAttrs>& attrs_, uint8_t mask_)
+        : SubgraphKey(attrs_), mask(mask_) {}
 
     size_t hash() const;
-    bool operator==(const SnippetKey& rhs) const;
+    bool operator==(const SubgraphShapeAgnosticKey& rhs) const;
+
+    uint8_t mask = 0;
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp]: Yeah, good point, thanks

File: src/plugins/intel_cpu/src/transformations/snippets/x64/pass/snippets_mark_skipped.cpp [hunk 18]
```diff
@@ -200,10 +200,7 @@ bool isSuitableMatMulParent(const std::shared_ptr<const Node> &node) {
 }
 // From Reduce::canFuse() corner case. CanFuseSimpleOperation is covered by Misc
 inline bool isSuitableReduceParent(const std::shared_ptr<const Node> &node) {
-    bool is_suitable_reduce = ov::is_type<ov::op::util::ArithmeticReductionKeepDims>(node) && isSuitableMiscParent(node);
-    bool is_not_min_max = !ov::is_type<ov::op::v1::ReduceMax>(node) && !ov::is_type<ov::op::v1::ReduceMin>(node);
-    bool out_is_f32 = node->get_output_element_type(0) == ov::element::f32;
-    return is_suitable_reduce && is_not_min_max && out_is_f32;
+return ov::is_type<ov::op::util::ArithmeticReductionKeepDims>(node) && isSuitableMiscParent(node);
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/transformations/snippets/x64/pass/snippets_mark_skipped.cpp]: Done, thanks

File: src/common/snippets/include/snippets/lowered/expression.hpp (hunk 5 above)
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/expression.hpp]: Removed for now

File: src/common/snippets/src/lowered/pass/set_load_store_scalar.cpp [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::SetLoadStoreScalar")
+    bool modified = false;
+    for (auto expr_it = begin; expr_it != end; ++expr_it) {
+        const auto& expr = *expr_it;
+        if (const auto load = ov::as_type_ptr<op::Load>(expr->get_node())) {
+            const auto& desc = expr->get_input_port_descriptor(0);
+            const auto& dim = desc->get_shape()[utils::get_input_dim_idx(desc->get_layout(), 0)];
+            OPENVINO_ASSERT(!utils::is_dynamic_value(dim), "SetLoadStoreScalar expects static shapes");
+            if (dim == 1) {
+                load->set_count(1);
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/set_load_store_scalar.cpp:29]: Why do we set count only for scalar operations? We still walked through the whole IR and checked the dimensions, is it just for `InsertBroadcastMove` to work correctly?

File: src/common/snippets/src/op/subgraph.cpp [hunk 20]
```diff
@@ -462,40 +522,43 @@ snippets::Schedule Subgraph::generate(const BlockedShapeVector& blocked_input_sh
...
-    if (linear_ir.get_config().perf_count_mode != lowered::PerfCountMode::Disabled) {
-        lowered::pass::InsertPerfCount perf_count_pass({});
-        perf_count_pass.run(linear_ir, linear_ir.cbegin(), linear_ir.cend());
+    auto linear_ir = *lowered::LinearIRBuilder().clone(m_linear_ir);
+
+    if (is_dynamic()) {
+        ov::snippets::lowered::pass::PassPipeline shape_dependent_pipeline;
+        shape_dependent_pipeline.register_pass<ov::snippets::lowered::pass::SetLoadStoreScalar>();
+        shape_dependent_pipeline.register_pass<ov::snippets::lowered::pass::InsertBroadcastMove>();
+        shape_dependent_pipeline.register_pass<ov::snippets::lowered::pass::LoadMoveBroadcastToBroadcastLoad>();
```
Comment (IvanNovoselov) [inline src/common/snippets/src/op/subgraph.cpp:543]: Since we are trying to minimize runtime overheads in dynaimc scenario,it looks like we should rewrite it in one pass.
Also it seems that we insert Brodacasts only after Loads now, no? In this case we can replace Load with BrodcastLoad straight away.

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 21]
```diff
@@ -51,117 +51,166 @@ std::mutex err_print_lock;
 #include "transformations/tpp/x64/pass/lowered/set_tpp_leading_dim.hpp"
 #endif
 
-using namespace dnnl::impl::utils;
-using namespace dnnl::impl::cpu;
-using namespace dnnl::impl::cpu::x64;
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp:56]: Minor: why did you remove the usings?
Now we have to spell `dnnl::impl::cpu::x64::mayiuse` which is longer and less readable

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 22]
```diff
@@ -379,85 +488,116 @@ void Snippet::initOptimalPrimitiveDescriptor() {
...
-ov::element::Type Snippet::getRuntimePrecision() const {
-    std::vector<ov::element::Type> inputPrecisions;
-    for (size_t i = 0; i < getParentEdges().size(); i++) {
-        auto parentEdge = getParentEdgeAt(i);
-        if (parentEdge && parentEdge->getStatus() == Edge::Status::Validated && !parentEdge->getParent()->isConstant()) {
-            inputPrecisions.emplace_back(DnnlExtensionUtils::DataTypeToElementType((parentEdge->getMemoryPtr()->getDataType())));
-        }
+uint8_t Subgraph::getBroadcastingMask(const std::vector<VectorDims>& input_shapes) {
+    uint8_t mask = 0;
+    // TODO: add check for non-eltwise inputs
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp:530]: We at least should check `has_domain_sensitive_ops` here

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 23]
```diff
@@ -379,85 +488,116 @@ void Snippet::initOptimalPrimitiveDescriptor() {
...
-        auto cache = context->getParamsCache();
-        auto result = cache->getOrCreate(key, builder);
-        execPtr = result.first;
-        if (!execPtr) {
-            OPENVINO_THROW("Executor is not created for node ", getName(), ".");
+void Subgraph::prepareParams() {
+    const auto cache = context->getParamsCache();
+
+    auto builder = [this, cache](const SubgraphKey& key) -> std::shared_ptr<SubgraphExecutor> {
+        const auto& cpu_config = ov::as_type_ptr<CPURuntimeConfig>(subgraph_attrs->snippet->update_runtime_config());
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp:569]: A bit confusing, `cpu_config` is more about things like isa
```suggestion
        const auto& snippet_config = ov::as_type_ptr<CPURuntimeConfig>(subgraph_attrs->snippet->update_runtime_config());
```

File: src/common/snippets/include/snippets/op/subgraph.hpp [hunk 24]
```diff
@@ -144,29 +134,42 @@ class Subgraph : public ov::op::util::SubGraphOp {
...
-                                   const std::vector<snippets::pass::Manager::PositionedPassBase>& = {});
+                                   const std::vector<snippets::pass::Manager::PositionedPassBase>& = {}) const;
+
+    void control_flow_transformations(size_t min_parallel_work_amount = 8, size_t min_kernel_work_amount = 256,
+                                      const std::shared_ptr<IShapeInferSnippetsFactory>& shape_infer_factory = std::make_shared<IShapeInferSnippetsFactory>(),
+                                      const std::shared_ptr<lowered::pass::PassConfig>& lowered_pass_config = std::make_shared<lowered::pass::PassConfig>(),
+                                      const std::vector<snippets::lowered::pass::PassPipeline::PositionedPassLowered>& lowered_backend_passes = {});
+
     std::shared_ptr<lowered::LinearIR>
     convert_body_to_linear_ir(size_t min_parallel_work_amount = 8, size_t min_kernel_work_amount = 256,
```
Comment (a-sidorova) [inline src/common/snippets/include/snippets/op/subgraph.hpp:145]: Will remove when the [PR](https://github.com/openvinotoolkit/openvino/pull/24152) is merged

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 21 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp]: After rebase on ARM PR - arm-specific jobs are failing on build stage, they cannot recognize `aarch64::mayuse()`. Also PR with ARM introduces the full namespaces there - so I left it so.

File: src/common/snippets/src/runtime_configurator.cpp [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+
+const std::shared_ptr<RuntimeConfig>& RuntimeConfigurator::get_updated_config(const std::shared_ptr<lowered::LinearIR>& linear_ir) {
+    // First initialization
+    if (m_io_num == 0) {
+        initialization(linear_ir);
+    }
+
+    update(linear_ir);
+    update_latest_shapes();
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/runtime_configurator.cpp:27]: Discussed offline: no need now so we can make simpler method, moved to `update()`.

**Review by v-Golubev — APPROVED** [review]
Great work! 🔥

File: src/common/snippets/src/lowered/loop_info.cpp [hunk 26]
```diff
@@ -350,6 +350,16 @@ const std::vector<int64_t>& ExpandedLoopInfo::get_data_sizes() const {
     return m_data_sizes;
 }
 
+void ExpandedLoopInfo::update_ptr_increments(const std::vector<int64_t>& new_values) {
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/loop_info.cpp:353]: When we discussed [this comment](https://github.com/openvinotoolkit/openvino/pull/21922#discussion_r1601375422) my idea was to employ move semantics. So the vectors should be passed by value, not by const reference. Then (as we discussed offline) compiler will be able to create more efficient code without redundant memory allocation.

File: src/common/snippets/include/snippets/lowered/loop_info.hpp [hunk 27]
```diff
@@ -404,6 +404,29 @@ class ExpandedLoopInfo : public LoopInfo {
...
+     * @brief Update `m_ptr_increments` using copy values from `new_values`.
+     *        The count of new values must be equal to the count of current increments.
+     * @param new_values vector of new pointer increments
+     */
+    void update_ptr_increments(const std::vector<int64_t>& new_values);
+    /**
+     * @brief Update `m_ptr_increments` setting `new_value` to all current values.
+     * @param new_value new value of all pointer increments
+     */
+    void update_ptr_increments(int64_t new_value);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/loop_info.hpp]: Left a follow-up comment: we should take advantage of move semantics here

File: src/common/snippets/src/lowered/pass/set_load_store_scalar.cpp (hunk 19 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/pass/set_load_store_scalar.cpp:29]: Ok. let's discuss later.

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 22 above)
Comment (IvanNovoselov) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp:694]: Works for now, but we should think about creating a set of predefined "special" masks in future

**Review by IvanNovoselov — APPROVED** [review]
Big step towards dynamism. It wasn't easy, but we made it 😁 🎉
Well done! 🥇
